R/rows.R

```diff
@@ -0,0 +1,610 @@
+#' @export
+#' @inheritParams dplyr::rows_insert
+#' @param conflict How should keys in `y` that already exist in `x` be handled?
+#'   One of:
+#'   - `"error"`, the default, will error if there are any keys in `y` that
+#'     already exist in `x`. This requires a corresponding unique constraint in
+#'     `x`.
+#'   - `"ignore"` will not insert rows in `y` with keys that already exist in `y`.
+#' @param returning Columns to return.
+#'
+#' @importFrom dplyr rows_insert
+#' @rdname rows-db
+rows_insert.tbl_lazy <- function(x,
+                                 y,
+                                 by = NULL,
+                                 ...,
+                                 conflict = c("error", "ignore"),
+                                 copy = FALSE,
+                                 in_place = FALSE,
+                                 returning = NULL) {
+  check_dots_empty()
+  rows_check_in_place(x, in_place)
+  name <- target_table_name(x, in_place)
+
+  conflict <- rows_check_conflict(conflict)
+
+  y <- auto_copy(x, y, copy = copy)
+
+  rows_check_containment(x, y)
+
+  by <- rows_check_by(by, y)
+
+  rows_check_key(x, by, "x")
+  rows_check_key(y, by, "y", unique = TRUE)
+
+  new_columns <- setdiff(colnames(y), by)
+
+  returning_cols <- rows_check_returning(x, returning, enexpr(returning))
+
+  if (!is_null(name)) {
+    if (conflict == "error") {
+      inform("Duplicated rows only produce an error if there is a unique constraint on `x`.")
+    }
+
+    sql <- sql_query_insert(
+      con = remote_con(x),
+      x_name = name,
+      y = y,
+      by = by,
+      ...,
+      conflict = conflict,
+      returning_cols = returning_cols
+    )
+
+    rows_get_or_execute(x, sql, returning_cols)
+  } else {
+    if (conflict == "error") {
+      abort('`conflict = "error"` is not supported for `in_place = FALSE`.')
```

Comment:
We still could run a query like `union_all() %>% count() %>% filter(n > 1) %>% head(1) %>% collect()` and error if we see at least one row here, along with a message that this is expensive and users can use `conflict = "ignore"` (meaning that conflicts are allowed). The alternative `conflict = "drop"` can also be expensive on the database.

Reply:
I think it is worth supporting `conflict = "error"` but `collecting()` by default does not sound good to me. So, maybe `conflict` should not be an optional argument for `rows_insert.tbl_lazy(in_place = FALSE)`.

Reply:
We pay extra for the checks in dplyr too, and we collect at most one row. We could give a message if the `conflict` argument is missing, to ensure that users are aware?